<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from marvin.urdf.xacro              | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<!-- Open_Manipulator Chain -->
<robot name="marvin">
  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="white">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>
  <material name="red">
    <color rgba="0.8 0.0 0.0 1.0"/>
  </material>
  <material name="blue">
    <color rgba="0.0 0.0 0.8 1.0"/>
  </material>
  <material name="green">
    <color rgba="0.0 0.8 0.0 1.0"/>
  </material>
  <material name="grey">
    <color rgba="0.5 0.5 0.5 1.0"/>
  </material>
  <material name="orange">
    <color rgba="1.0 0.4235294117647059 0.0392156862745098 1.0"/>
  </material>
  <material name="brown">
    <color rgba="0.8705882352941177 0.8117647058823529 0.7647058823529411 1.0"/>
  </material>
  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
  <!--xacro:include filename="/home/jaeho/marvin/ros2_ws/src/marvin/urdf2/open_manipulator.gazebo.xacro" /-->
  <transmission name="tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint1">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint2">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor2">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint3">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor3">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint4">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor4">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran5">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="gripper">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor5">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran6">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="gripper_sub">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor6">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!-- World -->
  <link name="world"/>
  <!-- World fixed joint-->
  <joint name="world_fixed" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="world"/>
    <child link="torso"/>
  </joint>
  <!-- Torso Link -->
  <link name="torso">
    <collision name="Cube">
      <origin rpy="0.00000 0.00000 0.00000" xyz="0.00000 0.00000 -0.22261"/>
      <geometry>
        <box size="0.16000 0.06000 0.60000"/>
      </geometry>
    </collision>
    <visual name="Cube">
      <origin rpy="0.00000 0.00000 0.00000" xyz="0.00000 0.00000 -0.22261"/>
      <geometry>
        <box size="0.16000 0.06000 0.60000"/>
      </geometry>
    </visual>
    <parent link="world"/>
  </link>
  <joint name="left_arm_joint" type="fixed">
    <parent link="torso"/>
    <child link="link1_left"/>
    <origin rpy="0 1.5708 0" xyz="0.08 0 0"/>
  </joint>
  <joint name="right_arm_joint" type="fixed">
    <parent link="torso"/>
    <child link="link1_right"/>
    <origin rpy="0 1.5708 3.1415926535" xyz="-0.08 0 0"/>
    <!-- Rotate by pi -->
  </joint>
  <!-- Link 1 -->
  <link name="link1_left">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="3.0876154e-04 0.0000000e+00 -1.2176461e-04"/>
      <mass value="7.9119962e-02"/>
      <inertia ixx="1.2505234e-05" ixy="0.0" ixz="-1.7855208e-07" iyy="2.1898364e-05" iyz="0.0" izz="1.9267361e-05"/>
    </inertial>
    <parent link="left_arm_joint"/>
  </link>
  <!-- Joint 1 -->
  <joint name="joint1_left" type="revolute">
    <parent link="link1_left"/>
    <child link="link2_left"/>
    <origin rpy="0 0 0" xyz="0.012 0.0 0.017"/>
    <axis xyz="0 0 1"/>
    <limit effort="1" lower="-2.827433388230814" upper="2.827433388230814" velocity="4.8"/>
  </joint>
  <!--  Link 2 -->
  <link name="link2_left">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.019"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.019"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="-3.0184870e-04 5.4043684e-04 0.047433463999999995"/>
      <mass value="9.8406837e-02"/>
      <inertia ixx="3.4543422e-05" ixy="-1.6031095e-08" ixz="-3.8375155e-07" iyy="3.2689329e-05" iyz="2.8511935e-08" izz="1.8850320e-05"/>
    </inertial>
  </link>
  <!--  Joint 2 -->
  <joint name="joint2_left" type="revolute">
    <parent link="link2_left"/>
    <child link="link3_left"/>
    <origin rpy="0 0 0" xyz="0.0 0.0 0.0595"/>
    <axis xyz="0 1 0"/>
    <limit effort="1" lower="-1.790707812546182" upper="1.5707963267948966" velocity="4.8"/>
  </joint>
  <!--  Link 3 -->
  <link name="link3_left">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="1.0308393e-02 3.7743363e-04 1.0170197e-01"/>
      <mass value="1.3850917e-01"/>
      <inertia ixx="3.3055381e-04" ixy="-9.7940978e-08" ixz="-3.8505711e-05" iyy="3.4290447e-04" iyz="-1.5717516e-06" izz="6.0346498e-05"/>
    </inertial>
  </link>
  <!--  Joint 3 -->
  <joint name="joint3_left" type="revolute">
    <parent link="link3_left"/>
    <child link="link4_left"/>
    <origin rpy="0 0 0" xyz="0.024 0 0.128"/>
    <axis xyz="0 1 0"/>
    <limit effort="1" lower="-0.9424777960769379" upper="1.382300767579509" velocity="4.8"/>
  </joint>
  <!--  Link 4 -->
  <link name="link4_left">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link4.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link4.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="9.0909590e-02 3.8929816e-04 2.2413279e-04"/>
      <mass value="1.3274562e-01"/>
      <inertia ixx="3.0654178e-05" ixy="-1.2764155e-06" ixz="-2.6874417e-07" iyy="2.4230292e-04" iyz="1.1559550e-08" izz="2.5155057e-04"/>
    </inertial>
  </link>
  <!--  Joint 4 -->
  <joint name="joint4_left" type="revolute">
    <parent link="link4_left"/>
    <child link="link5_left"/>
    <origin rpy="0 0 0" xyz="0.124 0.0 0.0"/>
    <axis xyz="0 1 0"/>
    <limit effort="1" lower="-1.790707812546182" upper="2.0420352248333655" velocity="4.8"/>
  </joint>
  <!--  Link 5 -->
  <link name="link5_left">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="4.4206755e-02 3.6839985e-07 8.9142216e-03"/>
      <mass value="1.4327573e-01"/>
      <inertia ixx="8.0870749e-05" ixy="0.0" ixz="-1.0157896e-06" iyy="7.5980465e-05" iyz="0.0" izz="9.3127351e-05"/>
    </inertial>
  </link>
  <!--  Gripper joint -->
  <joint name="gripper_left" type="prismatic">
    <parent link="link5_left"/>
    <child link="gripper_link_left"/>
    <origin rpy="0 0 0" xyz="0.0817 0.021 0.0"/>
    <axis xyz="0 1 0"/>
    <limit effort="1" lower="-0.010" upper="0.019" velocity="4.8"/>
  </joint>
  <!--  Gripper link -->
  <link name="gripper_link_left">
    <visual>
      <origin rpy="0 0 0" xyz="0.0 0.0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link_grip_l.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0.0 0.0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link_grip_l.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="1.0e-03"/>
      <inertia ixx="1.0e-03" ixy="0.0" ixz="0.0" iyy="1.0e-03" iyz="0.0" izz="1.0e-03"/>
    </inertial>
  </link>
  <!--  Gripper joint sub -->
  <joint name="gripper_sub_left" type="prismatic">
    <parent link="link5_left"/>
    <child link="gripper_link_sub_left"/>
    <origin rpy="0 0 0" xyz="0.0817 -0.021 0"/>
    <axis xyz="0 -1 0"/>
    <limit effort="1" lower="-0.010" upper="0.019" velocity="4.8"/>
    <mimic joint="gripper_left" multiplier="1"/>
  </joint>
  <!--  Gripper link sub -->
  <link name="gripper_link_sub_left">
    <visual>
      <origin rpy="0 0 0" xyz="0.0 -0.0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link_grip_r.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0.0 -0.0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link_grip_r.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="1.0e-03"/>
      <inertia ixx="1.0e-03" ixy="0.0" ixz="0.0" iyy="1.0e-03" iyz="0.0" izz="1.0e-03"/>
    </inertial>
  </link>
  <!-- end effector joint -->
  <joint name="end_effector_joint_left" type="fixed">
    <origin rpy="0 0 0" xyz="0.126 0.0 0.0"/>
    <parent link="link5_left"/>
    <child link="end_effector_link_left"/>
  </joint>
  <!-- end effector link -->
  <link name="end_effector_link_left">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="1.0e-03"/>
      <inertia ixx="1.0e-03" ixy="0.0" ixz="0.0" iyy="1.0e-03" iyz="0.0" izz="1.0e-03"/>
    </inertial>
  </link>
  <!-- Left Arm -->
  <!-- Link 1 -->
  <link name="link1_right">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="3.0876154e-04 0.0000000e+00 -1.2176461e-04"/>
      <mass value="7.9119962e-02"/>
      <inertia ixx="1.2505234e-05" ixy="0.0" ixz="-1.7855208e-07" iyy="2.1898364e-05" iyz="0.0" izz="1.9267361e-05"/>
    </inertial>
    <parent link="right_arm_joint"/>
  </link>
  <!-- Joint 1 -->
  <joint name="joint1_right" type="revolute">
    <parent link="link1_right"/>
    <child link="link2_right"/>
    <origin rpy="0 0 0" xyz="0.012 0.0 0.017"/>
    <axis xyz="0 0 1"/>
    <limit effort="1" lower="-2.827433388230814" upper="2.827433388230814" velocity="4.8"/>
  </joint>
  <!--  Link 2 -->
  <link name="link2_right">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.019"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.019"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="-3.0184870e-04 5.4043684e-04 0.047433463999999995"/>
      <mass value="9.8406837e-02"/>
      <inertia ixx="3.4543422e-05" ixy="-1.6031095e-08" ixz="-3.8375155e-07" iyy="3.2689329e-05" iyz="2.8511935e-08" izz="1.8850320e-05"/>
    </inertial>
  </link>
  <!--  Joint 2 -->
  <joint name="joint2_right" type="revolute">
    <parent link="link2_right"/>
    <child link="link3_right"/>
    <origin rpy="0 0 0" xyz="0.0 0.0 0.0595"/>
    <axis xyz="0 1 0"/>
    <limit effort="1" lower="-1.790707812546182" upper="1.5707963267948966" velocity="4.8"/>
  </joint>
  <!--  Link 3 -->
  <link name="link3_right">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="1.0308393e-02 3.7743363e-04 1.0170197e-01"/>
      <mass value="1.3850917e-01"/>
      <inertia ixx="3.3055381e-04" ixy="-9.7940978e-08" ixz="-3.8505711e-05" iyy="3.4290447e-04" iyz="-1.5717516e-06" izz="6.0346498e-05"/>
    </inertial>
  </link>
  <!--  Joint 3 -->
  <joint name="joint3_right" type="revolute">
    <parent link="link3_right"/>
    <child link="link4_right"/>
    <origin rpy="0 0 0" xyz="0.024 0 0.128"/>
    <axis xyz="0 1 0"/>
    <limit effort="1" lower="-0.9424777960769379" upper="1.382300767579509" velocity="4.8"/>
  </joint>
  <!--  Link 4 -->
  <link name="link4_right">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link4.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link4.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="9.0909590e-02 3.8929816e-04 2.2413279e-04"/>
      <mass value="1.3274562e-01"/>
      <inertia ixx="3.0654178e-05" ixy="-1.2764155e-06" ixz="-2.6874417e-07" iyy="2.4230292e-04" iyz="1.1559550e-08" izz="2.5155057e-04"/>
    </inertial>
  </link>
  <!--  Joint 4 -->
  <joint name="joint4_right" type="revolute">
    <parent link="link4_right"/>
    <child link="link5_right"/>
    <origin rpy="0 0 0" xyz="0.124 0.0 0.0"/>
    <axis xyz="0 1 0"/>
    <limit effort="1" lower="-1.790707812546182" upper="2.0420352248333655" velocity="4.8"/>
  </joint>
  <!--  Link 5 -->
  <link name="link5_right">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="4.4206755e-02 3.6839985e-07 8.9142216e-03"/>
      <mass value="1.4327573e-01"/>
      <inertia ixx="8.0870749e-05" ixy="0.0" ixz="-1.0157896e-06" iyy="7.5980465e-05" iyz="0.0" izz="9.3127351e-05"/>
    </inertial>
  </link>
  <!--  Gripper joint -->
  <joint name="gripper_right" type="prismatic">
    <parent link="link5_right"/>
    <child link="gripper_link_right"/>
    <origin rpy="0 0 0" xyz="0.0817 0.021 0.0"/>
    <axis xyz="0 1 0"/>
    <limit effort="1" lower="-0.010" upper="0.019" velocity="4.8"/>
  </joint>
  <!--  Gripper link -->
  <link name="gripper_link_right">
    <visual>
      <origin rpy="0 0 0" xyz="0.0 0.0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link_grip_l.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0.0 0.0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link_grip_l.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="1.0e-03"/>
      <inertia ixx="1.0e-03" ixy="0.0" ixz="0.0" iyy="1.0e-03" iyz="0.0" izz="1.0e-03"/>
    </inertial>
  </link>
  <!--  Gripper joint sub -->
  <joint name="gripper_sub_right" type="prismatic">
    <parent link="link5_right"/>
    <child link="gripper_link_sub_right"/>
    <origin rpy="0 0 0" xyz="0.0817 -0.021 0"/>
    <axis xyz="0 -1 0"/>
    <limit effort="1" lower="-0.010" upper="0.019" velocity="4.8"/>
    <mimic joint="gripper_right" multiplier="1"/>
  </joint>
  <!--  Gripper link sub -->
  <link name="gripper_link_sub_right">
    <visual>
      <origin rpy="0 0 0" xyz="0.0 -0.0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link_grip_r.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0.0 -0.0 0"/>
      <geometry>
        <mesh filename="../meshes/stl/chain_link_grip_r.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="1.0e-03"/>
      <inertia ixx="1.0e-03" ixy="0.0" ixz="0.0" iyy="1.0e-03" iyz="0.0" izz="1.0e-03"/>
    </inertial>
  </link>
  <!-- end effector joint -->
  <joint name="end_effector_joint_right" type="fixed">
    <origin rpy="0 0 0" xyz="0.126 0.0 0.0"/>
    <parent link="link5_right"/>
    <child link="end_effector_link_right"/>
  </joint>
  <!-- end effector link -->
  <link name="end_effector_link_right">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="1.0e-03"/>
      <inertia ixx="1.0e-03" ixy="0.0" ixz="0.0" iyy="1.0e-03" iyz="0.0" izz="1.0e-03"/>
    </inertial>
  </link>
  <!-- Right Arm (Mirrored) -->
</robot>
